### PR TITLE
Apple2: Implement statvfs() and stat() 

### PR DESCRIPTION
**Pull request by colinleroy**
Edit of the description: following below comments, this PR is now implementing stat() and statvfs() for the APPLE2 platform.

### CODE REVIEW

**Comment by oliverschmidt** [comment]
Hi,

First of all sorry for the delayed response, but the time before xmas way (too) busy.

I - of course - like the idea to have this functionality made available. However, I must admit that I'm rather unhappy about the way it is made available...

The DIO API is about "media". I refer to this term in the context of this layering - ordered from low layer to high layer:

1. Device/Drive
2. Media
3. File System

It's not about "device" as things like sector count (and even sector size) can be different for different media types which work in the same drive. But it's also not about file system as sectors are in general below the file system layer.

What might be confusing is that the dio_query_sectcount() implementation on the Apple II uses the ProDOS file system. But this is just an implementation detail. I've done it this way because any other approach would have caused quite some issues without providing tangible benefits.

But in contrast to the sector count, the number of free sectors is totally clearly a file system property - and therefore doesn't belong to DIO.

Instead of inventing a new function, I suggest to implement statvfs(). My proposal is to set all fields of `struct statvfs` to zero but:

* f_bsize: $256
* f_frsize: $256
* f_blocks: same as dio_query_sectcount()
* f_bfree: "the" thing
* f_namemax: same as FILENAME_MAX

Furthermore I propose to deviate from https://pubs.opengroup.org/onlinepubs/009695299/functions/statvfs.html in that `path` doesn't work for any file name on a volume (which might be relative and would require fiddling with PREFIX then etc.), but only for the volume name itself with a prepended slash as it is i.e. returned from getdevicedir(). Or - if you prefer - for any file name as long as it is absolute.

The idea is to call `ON_LINE` with `unit_num` set to 0 and compare the results with (a part of) `path`.

For course this is way more heavyweight than your original idea. But beside being a standard it also has the benefit that a program like the one that made you implement the feature doesn't need to know anything about volumes / units or alike.

In case the user want's to get an overview on the free sectors on all volumes - or a program seeks for a volume with certain amount of free sectors, the approach would be combine those functions:

1. getfirstdevice()/getnextdevice()
2. getdevicedir()
3. statvfs()

Regards,
Oliver

**Comment by colinleroy** [comment]
Hi Oliver!
seems like a very good idea, thanks!

**Comment by oliverschmidt** [comment]
> I didn't add ifdefs around the .h declaration because the other dio_* functions don't have them even though they're not implemented for each platform.

That's totally normal for all kinds of functions. It's up to the cc65 user to interpret the linker error accordingly.

BTW: If you haven't signed up for https://cc65.github.io/mailing-lists.html#cc65-devel yet I'd suggest that you do so and - when this PR is merged - post there about you having contributed a statvfs() for the Apple II and that others might want to do the same for e.g. CBM or Atari machines.

**Comment by colinleroy** [comment]
Will do! Thanks.
what do you think about implementing stat() too while I'm at it, or at least a subset of it?

**Comment by colinleroy** [comment]
Another question, would it be good to use iobuf_alloc/free instead of malloc/free for those and dio_querysectcount? They'd return too big buffers, but users of iobuf0800 wouldn't have to worry about extra allocation?
or would you rather not?

**Comment by oliverschmidt** [comment]
I haven't looked at your actual free space logic yet but I want to make sure that you are aware of the following statement regarding the GET_FILE_INFO call:

_When file information about a volume directory is requested, the total number of blocks on the volume is returned in the aux_type field and the total blocks for all files is returned in blocks_used._

**Comment by colinleroy** [comment]
> I haven't looked at your actual free space logic yet but I want to make sure that you are aware of the following statement in the ProDOS TechRefMan regarding the GET_FILE_INFO call:
> 
> _When file information about a volume directory is requested, the total number of blocks on the volume is returned in the aux_type field and the total blocks for all files is returned in blocks_used._

I was not, and it seems so much easier (and faster) than iterating over the bits of the bytes of the bitmap blocks 😅

**Comment by colinleroy** [comment]
By the way, this thing about GET_FILE_INFO makes the ON_LINE (unitnum=0) call you mention in your first comment unnecessary, doesn't it? Using GET_FILE_INFO will allow implementing both statvfs and stat (with an extra GET_EOF call to get a file's precise size in bytes) easy.

**Comment by oliverschmidt** [comment]
> By the way, this thing about GET_FILE_INFO makes the ON_LINE (unitnum=0) call you mention in your first comment unnecessary, doesn't it?

You're of course right. I got things mixed up. My proposal to have statvfs() either take a slash-prefixed volume name (or an absolute path containing such a slash-prefixed volume name) fits perfectly the GET_FILE_INFO parameter. At some point I knew that but at some point I "lost>" it :confused:

And in case you wonder: Yes, it might make sense to re-implement dio_query_sectcount() based on getdevicedir() and GET_FILE_INFO. I leave it up to you if you want to try that or leave dio_query_sectcount() alone.

> Using GET_FILE_INFO will allow implementing both statvfs and stat (with an extra GET_EOF call to get a file's precise size in bytes) easy, without even bothering with the device number.

Yes, I'd appreciate a stat() implementation based on GET_FILE_INFO and GET_EOF. However, please be aware that GET_EOF requires to open/close the file. Nevertheless it seems that the byte size is worth it.

**Comment by colinleroy** [comment]
> By the way, this thing about GET_FILE_INFO makes the ON_LINE (unitnum=0) call you mention in your first comment unnecessary, doesn't it? Using GET_FILE_INFO will allow implementing both statvfs and stat (with an extra GET_EOF call to get a file's precise size in bytes) easy.

Thanks for confirmations!
concerning stat(), do you think I should stick to the posix definition of struct stat or take liberties with it, as is done with dirent? I fear doing mktime for the crezte/modify dates is going to bloat code size quite a bit.

**Comment by colinleroy** [comment]
> > By the way, this thing about GET_FILE_INFO makes the ON_LINE (unitnum=0) call you mention in your first comment unnecessary, doesn't it?
> 
> You're of course right. I got things mixed up. My proposal to have statvfs() either take a slash-prefixed volume name (or an absolute path containing such a slash-prefixed volume name) fits perfectly the GET_FILE_INFO parameter. At some point I knew that but at some point I "lost>" it 😕
> 
> And in case you wonder: Yes, it might make sense to re-implement dio_query_sectcount() based on getdevicedir() and GET_FILE_INFO. I leave it up to you if you want to try that or leave dio_query_sectcount() alone.
> 
> > Using GET_FILE_INFO will allow implementing both statvfs and stat (with an extra GET_EOF call to get a file's precise size in bytes) easy, without even bothering with the device number.
> 
> Yes, I'd appreciate a stat() implementation based on GET_FILE_INFO and GET_EOF. However, please be aware that GET_EOF requires to open/close the file. Nevertheless it seems that the byte size is worth it.

Thanks for the confirmations!
concerning stat(), do you think I should stick to the posix definition of struct stat or take liberties with it, as is done with dirent? I fear doing mktime for the crezte/modify dates is going to bloat code size quite a bit.

**Comment by oliverschmidt** [comment]
I was in fact already about to write something about `struct dirent` :wink:

Yes, I propose to have custom fields:
* `st_type` (identical to `d_type`)
* `st_auxtype` (identical to `d_auxtype`)
* `st_access` (identical to `d_access`)

However, the situation regarding the time format isn't comparable to the `struct dirent`. https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/dirent.h.html describes only `d_ino` and `d_name`. Everything else is proprietary. In contrast https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_stat.h.html clearly defines the timestamps to be of type `struct timespec`. It seems to me we don't have much room for maneuver here...

**Comment by oliverschmidt** [comment]
> Another question, would it be good to use iobuf_alloc/free instead of malloc/free for those and dio_querysectcount, opendir and friends? They'd return too big buffers, but users of iobuf0800 wouldn't have to worry about extra allocation?
or would you rather not?

Thanks for the proposal :smile:

I introduced the iobuf mechanism primarily because the alignment requirements and the size of the posix_memalign() code itself (written in C) added a LOT of overhead. I've thought about the pros and cons of that approach for some time and it seems to me we should rather stick with malloc/free.

**Comment by colinleroy** [comment]
Hi @oliverschmidt, Here you are :)

The stat() function weighs quite a bit, and pulls in a lot of things (mli, open, close, mktime, etc etc). In an otherwise empty program, linking stat() in costs 4.1kB. (429 bytes if open/malloc/free/mktime are already linked in). The time conversion are heavy...

statvfs is much lighter, 523 bytes.

File: libsrc/apple2/filename.s
Comment (oliverschmidt) [inline libsrc/apple2/filename.s:97]: Please use FILENAME_MAX here too.

File: asminc/apple2.inc
Comment (oliverschmidt) [inline asminc/apple2.inc:90]: * I think FILENAME_MAX doesn't belong here. This file is about Apple II hardware attributes, not about Apple II software attributes. It fits much better in libsrc/apple2/mli.inc.
* The definition should use `=`, not `:=` as it doesn't denote an address.
* Given that the C definition is `64+1` I'm wondering if it might be preferrable to use `64+1` here too.

File: doc/funcref.sgml
Comment (oliverschmidt) [inline doc/funcref.sgml:7148]: I suggest to use lowercase for `DISK` too.

File: include/sys/stat.h
Comment (oliverschmidt) [inline include/sys/stat.h:59]: Please remove two empty lines.

File: include/sys/stat.h
Comment (oliverschmidt) [inline include/sys/stat.h:70]: Please remove one empty line.

File: libsrc/apple2/mli.inc
Comment (oliverschmidt) [inline libsrc/apple2/mli.inc:87]: Thanks 👍🏻

File: libsrc/apple2/stat.s
Comment (oliverschmidt) [inline libsrc/apple2/stat.s:87]: It's amazing how well you adapted to the existing coding style !!!

Just one minor thing here: When two "actual" values are added, then the `+` is separated with blanks. But if it's "just the usual + 1" (especially for 16-bit high-byte related stuff), then the blanks are omitted.

File: libsrc/apple2/stat.s
Comment (oliverschmidt) [inline libsrc/apple2/stat.s:138]: Blanks (see above).

File: libsrc/apple2/stat.s
Comment (oliverschmidt) [inline libsrc/apple2/stat.s:141]: Blanks (see above).

File: libsrc/apple2/statvfs.s
Comment (oliverschmidt) [inline libsrc/apple2/statvfs.s:79]: Blanks (see above).

File: libsrc/apple2/statvfs.s
Comment (oliverschmidt) [inline libsrc/apple2/statvfs.s:92]: Blanks (see above).

File: libsrc/apple2/statvfs.s
Comment (oliverschmidt) [inline libsrc/apple2/statvfs.s:93]: Blanks (see above).

File: libsrc/apple2/statvfs.s
Comment (oliverschmidt) [inline libsrc/apple2/statvfs.s:96]: Blanks (see above).

File: libsrc/apple2/statvfs.s
Comment (oliverschmidt) [inline libsrc/apple2/statvfs.s:111]: Blanks (see above).

File: libsrc/apple2/statvfs.s
Comment (oliverschmidt) [inline libsrc/apple2/statvfs.s:116]: I believe that this value needs to end up in one way or another being `65`, not `64`.

**Review by oliverschmidt — CHANGES_REQUESTED** [review]

**Comment by oliverschmidt** [comment]
> In an otherwise empty program, linking stat() in costs 4.1kB.

That's indeed a lot :-(

I'm wondering if we could/should replace
```
struct stat {
    dev_t             st_dev;
    ino_t             st_ino;
    mode_t            st_mode;
    nlink_t           st_nlink;
    uid_t             st_uid;
    gid_t             st_gid;
    off_t             st_size;
    struct timespec   st_atim;
    struct timespec   st_ctim;
    struct timespec   st_mtim;
    #ifdef __APPLE2__
    unsigned char     st_access;
    unsigned char     st_type;
    unsigned int      st_auxtype;
    #endif
};
```
with
```
struct stat {
    dev_t             st_dev;
    ino_t             st_ino;
    mode_t            st_mode;
    nlink_t           st_nlink;
    uid_t             st_uid;
    gid_t             st_gid;
    off_t             st_size;
    struct timespec   st_atim;
    struct timespec   st_ctim;
    struct timespec   st_mtim;
    #ifdef __APPLE2__
    struct {
        unsigned day  :5;
        unsigned mon  :4;
        unsigned year :7;
    }                 st_cdate;
    struct {
        unsigned char min;
        unsigned char hour;
    }                 st_ctime;
    struct {
        unsigned day  :5;
        unsigned mon  :4;
        unsigned year :7;
    }                 st_mdate;
    struct {
        unsigned char min;
        unsigned char hour;
    }                 st_mtime;
    unsigned char     st_access;
    unsigned char     st_type;
    unsigned int      st_auxtype;
    #endif
};
```

The idea is to keep
```
    struct timespec   st_atim;
    struct timespec   st_ctim;
    struct timespec   st_mtim;
```
in place to stay compliant but just not fill them. After all, `st_atim` can't be filled anyway. Instead fill proprietary fields.

@mrdudz: The point here is that mktime() is pretty large. And it's almost certain that the user doesn't benefit from a timespec    at all. He won't try to subtract file timestamps or alike. Rather he only wants to display them, which adds additional cost for localtime(). So do you consider the approach described above as acceptable/preferable?

Note1: It's likely desirable to have a struct declared for the ProDOS time/date representation and use it both above and in dirent.h instead of repeating identical definitions.

Note2: It might be desirable to add some proprietary function that converts such a ProDOS time/date representation (as it is exposed both be readdir() and stat()) into a `struct tm`.

**Comment by colinleroy** [comment]
@oliverschmidt, thanks for your review! I will address those.
Concerning stat()'s size, mktime is part of it, but not only. There's also the requirement to open() to get the file size, which pulls in malloc,free,posix_memalign etc.
But I think using prodos "extensions" is a good idea too, and that will indeed shrink stat() a bit.

**Comment by oliverschmidt** [comment]
> There's also the requirement to open() to get the file size, which pulls in malloc,free,posix_memalign etc.

I see your point. However, it's very likely that a program using stat() will also want to open() at least one file. So  all that code is pulled in anyhow. But that's not true at all for mktime(), especially as there are likely (many) use cases where stat() is only called to get the file size and the timestamps are completely ingored.

**Comment by colinleroy** [comment]
> > There's also the requirement to open() to get the file size, which pulls in malloc,free,posix_memalign etc.
> 
> I see your point. However, it's very likely that a program using stat() will also want to open() at least one file. So all that code is pulled in anyhow. But that's not true at all for mktime(), especially as there are likely (many) use cases where stat() is only called to get the file size and the timestamps are completely ingored.

I wasn't disagreeing, and am 100% aligned with your view - I was just managing your expectations :wink: 
Latest push fixes the coding style issues, and implements what you suggested. stat() now weighs 2392 bytes in an otherwise empty program, 314 bytes in a program already using open().

**Comment by colinleroy** [comment]
Added some cleanup for pushname/popname, the "ownership" of it was unclear, now it is.

Concerning dio_querysectcount(): it could be rewritten with something along the lines of
```
... prepare a buffer for the C-style filename
jsr get_devicedir
jsr pushax
jsr mli_info_call
lda mliparam + MLI::INFO::AUX_TYPE
ldx mliparam + MLI::INFO::AUX_TYPE+1
... cleanup C-style filename buffer
```
but I don't think it's worth it: the size gain on dio_querysectcount() would be completely cancelled by the pulling in of get_devicedir plus the buffer handling.

**Comment by colinleroy** [comment]
If/when you think it's good to merge, you can ping if you want me to squash commits to keep a clean history on the main branch :)

File: libsrc/apple2/mli_file_info.s
Comment (oliverschmidt) [inline libsrc/apple2/mli_file_info.s:54]: I don't see a technical reason why these two functions need to be in a single file. Therefore I ask you to separate them in order to have i.e. exec() callers not link the unused code of mli_file_info.

File: asminc/time.inc
Comment (oliverschmidt) [inline asminc/time.inc:69]: I don't like this section here. Maybe mli.inc or <do you have a better idea?>

File: libsrc/apple2/mli_file_info.s
Comment (colinleroy) [reply libsrc/apple2/mli_file_info.s:54]: Good idea!

File: libsrc/apple2/stat.s
Comment (colinleroy) [reply libsrc/apple2/stat.s]: That was the plan, but I forgot about it midway !

Comment (colinleroy) [reply libsrc/apple2/stat.s]: I've put st_storagetype and st_blocks instead of st_reserved, they could be useful to someone and there's no reason to hide them IMHO

File: include/time.h
Comment (colinleroy) [reply include/time.h:91]: I wanted to originally, but then we'd have to include <apple2.h> in stat.h, dirent.h etc, and <time.h> in apple2.h, that seemed more gross in the end?

File: asminc/time.inc
Comment (colinleroy) [reply asminc/time.inc:69]: Same as for the .h files: putting it elsewhere could be possible, but not in apple2.inc either, or we'd have to include it there, too. Maybe a specific file asminc/prodos_datetime.inc ?

File: include/time.h
Comment (oliverschmidt) [reply include/time.h:91]: 1. I don't think so. Headers including other headers seems more normal to me than putting "all into one" header :smirk: I propose to `#include <target.h>` where necessary.
2. https://github.com/cc65/cc65/blob/master/include/apple2.h#L153-L167 was **NO** struct type declaration. It was a variable declaration. You can't just remove it.

File: asminc/time.inc
Comment (oliverschmidt) [reply asminc/time.inc:69]: I don't agree that it is the same as for the .h files: The .h files include other .h files "where necessary", but .inc files never include other .inc files - at least afaik. It's up to the user to know what to include in which order.

Given that there's no code using this .struct at all I propose to "simply" go without. In stat.inc we just write:
```
.struct stat
        ...
        st_blocks         .word
        st_mod_date       .word
        st_mod_time       .word
        st_create_date    .word
        st_create_time    .word
```

File: include/apple2.h
Comment (oliverschmidt) [inline include/apple2.h:168]: As discussed, restore the declaration of the variable `_datetime`.

File: libsrc/apple2/open.s
Comment (oliverschmidt) [inline libsrc/apple2/open.s:251]: As discussed, please restore this to the original state.

File: libsrc/apple2/prodos_datetime_to_posix_tm.s
Comment (oliverschmidt) [inline libsrc/apple2/prodos_datetime_to_posix_tm.s:99]: You moved this from .BSS to .CODE - this seems undesirable to me.

File: libsrc/apple2/prodos_datetime_to_posix_tm.s
Comment (oliverschmidt) [inline libsrc/apple2/prodos_datetime_to_posix_tm.s:5]: I think it's preferable to have this function return a value to the "internal" struct tm, just like localtime() and gmtime() do.

**Comment by oliverschmidt** [comment]
Sorry for coming up with this so late :flushed: but...

I don't think that the `prodos_` prefix is desirable. We only support ProDOS on the Apple II and that won't chnage. And after all we call it `open()` and not `prodos_open()` :smirk:

So `prodos_datetime` should become `datetime`.

In addition, we rather prefer short, cryptic names over long, descriptive names. Therefore I propose:

`time_t __fastcall__ mktime_dt(const struct datetime* dt);`
`struct tm* __fastcall__ gmtime_dt(const struct datetime* dt);`

File: libsrc/apple2/prodos_datetime_to_posix_tm.s
Comment (colinleroy) [reply libsrc/apple2/prodos_datetime_to_posix_tm.s]: nice catch!

File: include/time.h
Comment (colinleroy) [reply include/time.h]: 1. Addressed :)
2. 3.: Indeed! I read that too fast.

File: include/sys/stat.h
Comment (oliverschmidt) [inline include/sys/stat.h:67]: Please move this to apple2.h too.

File: include/time.h
Comment (colinleroy) [reply include/time.h]: About the `extern struct datetime _datetime`: I have put it back. It references __datetime from open.s and could indeed be used by cc65 users; however, it's not used internally and not mentioned in the docs. 
The fact I've put it back using the struct datetime definition instead of the "one shot" definition means its fields changed, from `_datetime.createdate/_datetime.createtime` to `_datetime.date/_datetime.time`.
Is this acceptable?

Comment (oliverschmidt) [reply include/time.h]: Thanks! That's exactly how I envisioned it. I believe this incompatible change is to be accepted. I don't see it being used a lot because of the missing documentation. Maybe you would want to add some documentation in another change(?).

**Comment by oliverschmidt** [comment]
And finally for the **REAL** nitpicking :stuck_out_tongue_winking_eye:

* The cc65 coding style is to put the dereference `*` at the type, not at the variable. See my proposed prototypes above.
* You sometimes write 'POSIX', 'UNIX' and 'POSIX-style'. I haven't checked if there a clear distinction. If there is, then I'm of course fine. Otherweise it might be preferable to consistently use one term.

**Comment by colinleroy** [comment]
> And finally for the **REAL** nitpicking 😜

:sweat_smile: 

>     * The cc65 coding style is to put the dereference `*` at the type, not at the variable. See my proposed prototypes above.

Fixed! Sorry.
 
>     * You sometimes write 'POSIX', 'UNIX' and 'POSIX-style'. I haven't checked if there a clear distinction. If there is, then I'm of course fine. Otherweise it might be preferable to consistently use one term.

I'm unsure too. I sneakily reworded it to "more portable" to avoid hurting my brain with reference documentations, apart for the time_t which everybody knows is a "UNIX timestamp" :-)

File: include/apple2.h
Comment (oliverschmidt) [inline include/apple2.h:156]: Inside the "sections", we only have only linefeed.

**Comment by oliverschmidt** [comment]
> Fixed! Sorry.

Certainly no reason to say "sorry"! Please adjust the documentation too.

And I noticed that (parts of) the example code doesn't adhere to the "strange" cc65 C code style with the black in front of function parameters.

**Comment by colinleroy** [comment]
> > Fixed! Sorry.
> 
> Certainly no reason to say "sorry"! Please adjust the documentation too.
> 
> And I noticed that (parts of) the example code doesn't adhere to the "strange" cc65 C code style with the blank in front of function parameters.

Fixed (and in apple2.h too)

File: include/apple2.h
Comment (oliverschmidt) [inline include/apple2.h:177]: Is this a correct comment here? I think _datetime only applies to open and fopen.

Comment (colinleroy) [reply include/apple2.h:177]: Uh indeed, that's a leftover from when I had put it on the struct.

**Comment by oliverschmidt** [comment]
Believe it or not :grin: I've nothing to complain about anymore :thumbsup:

Do you want to squash something or do you want me to merge as-is?

BTW: I've not checked anything regarding the actual functional correctness. But I have reason to believe that you'll take care of issues if they should show up :smile:

**Comment by colinleroy** [comment]
> Believe it or not 😁 I've nothing to complain about anymore 👍

Yay!

> Do you want to squash something or do you want me to merge as-is?

Squashing's done !

> BTW: I've not checked anything regarding the actual functional correctness. But I have reason to believe that you'll take care of issues if they should show up 😄

I have tested everything rather extensively, on volumes, directories, files, non-existent entities etc. And I will indeed be there in case of issues !

**Comment by colinleroy** [comment]
Oh wait just one info I want to add to the doc.

**Comment by oliverschmidt** [comment]
That was a "tiny bit" more work than you expected originally, I guess :stuck_out_tongue_winking_eye:

The branch name `apple2-dio-freesectcount` is a silent witness...

**Comment by colinleroy** [comment]
> That was a "tiny bit" more work than you expected originally, I guess 😜
> 
> The branch name `apple2-dio-freesectcount` is a silent witness...

Haha, yes, but it's better work :) 
Now I'm gonna remove my #ifdefs __CC65__ that were using stat() on linux and an opendir/readdir-based loop on 6502, cleanup my code, and regain a few hundred bytes, so I'm happy!

**Comment by oliverschmidt** [comment]
That reminds me on my work on Contiki. There too, it was a ping-pong of adding capabilities I needed to cc65 and then adding their usage to Contiki. I.e. exec() was born by the need to run wget from the the web browser. I remember that approach being very rewarding - so I'm with you :grinning:

**Comment by colinleroy** [comment]
> That reminds me on my work on Contiki. There too, it was a ping-pong of adding capabilities I needed to cc65 and then adding their usage to Contiki. I.e. exec() was born by the need to run wget from the the web browser. I remember that approach being very rewarding - so I'm with you 😀

This commit was very satisfying - and makes for a 221 bytes gain on my binary!
https://github.com/colinleroy/a2tools/commit/a6a8cfcc0b8e261b2424cc6c5589c9acd669d7ef

**Pull request merged**
